[Console] Simplify simulation of user inputs in CommandTester

chalasr opened the pull request:
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | https://github.com/symfony/symfony-docs/pull/6623 |

After @javiereguiluz pointed it in #17470, I open this PR to simplify the simulation of user inputs for testing a Command.
It would be done by calling `CommandTester::setUserInputs()` with an array of inputs as argument, and so make the CommandTester creating an input stream from the inputs set by the developer, then call `QuestionHelper::setInputStream` and assign it to the helperSet of the command, sort as all is done automatically in one call.

Depends on #18999

chalasr committed:
[Console] Set user inputs from CommandTester

Rename getInputStream() to getStream(), taking inputs as argument

Make createStream static, typehint  as array

nicolas-grekas commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -129,4 +137,48 @@ public function getStatusCode()
+    {
+        $stream = fopen('php://memory', 'r+', false);
+
+        fputs($stream, implode("\n", $this->userInputs));
```
I think we use fwrite instead of fputs

chalasr replied:
I tried to follow this http://symfony.com/doc/current/components/console/helpers/questionhelper.html#testing-a-command-that-expects-input as possible, please let me know if you still think that I should use fwrite.

nicolas-grekas commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -64,6 +68,10 @@ public function execute(array $input, array $options = array())
             $input = array_merge(array('command' => $this->command->getName()), $input);
         }
 
+        if (0 < count($this->userInputs)) {
```
`if ($this->userInputs) {`

chalasr replied:
Fixed

OskarStark commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -129,4 +137,48 @@ public function getStatusCode()
     }
+
+    /**
+     * Sets the faked user inputs.
```
would remove the word `faked`

OskarStark commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -129,4 +137,48 @@ public function getStatusCode()
+    /**
+     * Gets the faked user inputs.
+     *
+     * @return array $inputs
```
only the type

OskarStark commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -129,4 +137,48 @@ public function getStatusCode()
+    }
+
+    /**
+     * Gets the faked user inputs.
```
same here

OskarStark commented:
this should be documented somewhere

thanks for your effort 👍

chalasr commented:
@OskarStark Thanks for the review.
I am waiting for the opinion of a maintainer before proposing the changes in the doc to make it in one shot.

javiereguiluz commented:
@chalasr could you please make the test more complex to simulate several inputs in a row? Thanks!

chalasr commented:
@javiereguiluz Of course, I added a failure test (3 questions for only 2 inputs) and make the first taking 3 questions for 3 inputs.

I also fixed the @OskarStark line notes and started to write documentation.

OskarStark commented:
> I also fixed the @OskarStark line notes and started to write documentation.

Thank you for your effort 👍

chalasr commented:
According to #17136, this will not work if the SymfonyStyle is used.
I'm working for find a clean solution, but making this SymfonyQuestionHelper out of the command HelperSet made interactive command testing very hard.
In a perfect world, the SymfonyQuestionHelper should be set in the helperSet and be accessible through `Command::getHelper('question')`. Actually the SymfonyStyle cannot access the Command and the Command cannot access the SymfonyStyle...

chalasr commented:
Doc [submitted](https://github.com/symfony/symfony-docs/pull/6623).

fabpot commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -129,4 +137,48 @@ public function getStatusCode()
+     *
+     * @return CommandTester
+     */
+    public function setUserInputs(array $inputs)
```
I would have named the method `setInputs()` instead.

fabpot commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -129,4 +137,48 @@ public function getStatusCode()
+     *
+     * @return array
+     */
+    public function getUserInputs()
```
Why do we need a getter?

chalasr replied:
We don't, I removed it.

chalasr commented:
I made the changes and updated the doc.

xabbuh commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -129,4 +137,48 @@ public function getStatusCode()
+     *
+     * @param array
+     *
+     * @return CommandTester
```
Imo this docblock as it is now is quite useless and could be removed as well (input types and the return type can be inferred by IDEs from the method itself) and the description doesn't add any value to the user apart from what the method name already states. Though explaining what `$inputs` exactly is expected to be would be something that justified its presence.

chalasr replied:
I added a description.

fabpot commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -129,4 +137,39 @@ public function getStatusCode()
+    /**
+     * Sets the user inputs.
+     *
+     * @param array A simple array of strings representing each input
```
`simple` should probably be removed here.

fabpot commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -129,4 +137,39 @@ public function getStatusCode()
+
+    /**
+     * Creates the command input stream.
+     */
```
the phpdoc should be removed.

fabpot commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -129,4 +137,39 @@ public function getStatusCode()
+    /**
+     * Creates the command input stream.
+     */
+    private function createInputStream()
```
I would prefer to have it return the stream instead of injecting it via `setInputStream()`, which can be done in the caller.

chalasr commented:
@fabpot I made the changes related to your line notes.

fabpot commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -64,6 +68,17 @@ public function execute(array $input, array $options = array())
+            }
+
+            $helper->setInputStream($this->createInputStream());
+        }
```
This looks wrong to me. Here we are only dealing with one specific case, which is the question helper. We should rather find a way to make this "generic" if possible.

chalasr replied:
@fabpot Only the QuestionHelper and SymfonyQuestionHelper have an `inputStream`, so I don't see a real benefit to make this looking for another object that the `question` helper.

I made this covering the case of SymfonyStyle is used (SymfonyQuestionHelper) in #18902, as this one is not part of the command helpset, we don't have any way to make it generic. It's [planned for 4.0 to move the setInputStream](https://github.com/symfony/symfony/issues/10844) method in the InputInterface, but ATM the QuesitonHelper is the only available place (and [documented] as is(http://symfony.com/doc/current/components/console/helpers/questionhelper.html#testing-a-command-that-expects-input)).
What do you think?

chalasr replied:
Adding an interface with a `setInputStream` method then implement it in the Input object could be a good solution. We could then automatically use `Input::$inputStream` in every helper that has an `inputStream`, working for both QuestionHelper and SymfonyQuestionHelper. At first looking for the `Input::$inputStream` while letting the `QuestionHelper::$inputStream` as deprecated for finally remove it in 4.0.
What do you think? It would close #18902 and ease interactive command testing properly.

chalasr replied:
@fabpot I just opened #18999 that would BTW solve this problem.
It gives:

```diff
if ($this->inputs) {
-     if (null === $this->command->getHelperSet()) {
-         $helper = new QuestionHelper();
-         $this->command->setHelperSet(new HelperSet(array($helper)));
-     } else {
-         $helper = $this->command->getHelper('question');
-     }
-     $helper->setInputStream($this->createInputStream());

+    $this->input->setStream($this->createInputStream());
}
```

chalasr commented:
Rebased this on #18999 to see the benefit by testing commands using both QuestionHelper and SymfonyQuestionHelper (via SymfonyStyle, no need of an HelperSet).

fabpot commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -39,46 +41,50 @@ public function __construct(Command $command)
-
-        return $this->statusCode = $this->command->run($this->input, $this->output);
-    }
+     /**
```
There is an indentation problem here for this method.

chalasr replied:
Fixed. The changes of this PR are separated in https://github.com/symfony/symfony/pull/18710/commits/69dd333eda69554690d586755e9706ffbd000d70

fabpot commented:
@chalasr Now that #18999 is merged, can you rebase this one?

fabpot commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -129,4 +135,29 @@ public function getStatusCode()
+        return $this;
+    }
+
+    private function getInputStream()
```
I would name it `createInputStream()` and pass `$this->inputs` as an argument.

fabpot replied:
I would then rename it to `createStream()`

fabpot commented:
Just noticed that the second commit messages contain all messages from squashed commits, can you remove them as they are useless? Thanks.

chalasr commented:
Made the changes, rebased and cleaned the commit message.

chalasr commented:
@fabpot As it takes the inputs as argument, I made the `getStream()` method static.

fabpot commented on src/Symfony/Component/Console/Tester/CommandTester.php:
```diff
@@ -129,4 +135,29 @@ public function getStatusCode()
+        return $this;
+    }
+
+    private static function createStream(array $inputs)
```
What's the benefit of marking it public? I would have kept it private.

ogizanagi replied:
@fabpot : Did you mean `static` ?
I guess it makes sense, as this method looks like a utility one.

fabpot replied:
No, I meant public. I think it should be private, I don't see why an end user would have to use this.

ogizanagi replied:
@fabpot : But the `createStream` method is actually `private`. Not `public`. 😕

chalasr replied:
@fabpot You surely misread it, it is public.

fabpot replied:
lol, sorry about the confusion.

fabpot commented:
Thank you @chalasr.
